src/sliceList.ml

```diff
@@ -0,0 +1,195 @@
+(* Double-linked list for burrow slices which acts as a higher-level interface for data stored in the AVL queue.
+ * This data structure allows for fast lookups of slices for a specific burrow, and
+ * functions which are adding or removing slices from the queue should use this module instead of
+ * AVL.ml directly since this module will automatically ensure that the burrow slice lists stay up
+ * to date.
+*)
+open Avl
+open Mem
+open LiquidationAuctionTypes
+open LiquidationAuctionPrimitiveTypes
+
+type slice_list_element = SliceListElement of (leaf_ptr * liquidation_slice)
+[@@deriving show]
+
+let slice_list_element_contents (e: slice_list_element) =
+  match e with
+  | SliceListElement (_, contents) -> contents.contents
+
+type slice_list_bounds = {
+  slice_list_youngest_ptr : leaf_ptr;
+  slice_list_oldest_ptr : leaf_ptr;
+}
+[@@deriving show]
+
+type slice_list_meta = {
+  slice_list_burrow: burrow_id;
+  slice_list_bounds: slice_list_bounds option
+}
+[@@deriving show]
+
+(* Question: Is it worth storing one of the end elements within the SliceList?  *)
+type slice_list = SliceList of slice_list_meta
+[@@deriving show]
+
+let slice_list_empty (burrow: burrow_id) = SliceList {slice_list_burrow=burrow; slice_list_bounds=(None:slice_list_bounds option);}
+
+let slice_list_is_empty (l: slice_list) =
+  let meta = match l with SliceList meta -> meta in
+  match meta.slice_list_bounds with
+  | Some _ -> false
+  | None -> true
+
+(* Constructs a burrow slice list for the given burrow id using the provided auction state *)
+let slice_list_from_auction_state (auctions: liquidation_auctions) (burrow_id: burrow_id) =
+  match Ligo.Big_map.find_opt burrow_id auctions.burrow_slices with
+  | None -> SliceList {slice_list_burrow=burrow_id; slice_list_bounds=(None:slice_list_bounds option)}
+  | Some bs ->
+    SliceList {
+      slice_list_burrow = burrow_id;
+      slice_list_bounds = Some {
+          slice_list_youngest_ptr = bs.youngest_slice;
+          slice_list_oldest_ptr = bs.oldest_slice;
+        };
+    }
+
+(* Constructs an element from a burrow leaf in the AVL *)
+let slice_list_from_leaf_ptr (auctions: liquidation_auctions) (ptr: leaf_ptr) =
+  let slice = avl_read_leaf auctions.avl_storage ptr in
+  let element = SliceListElement (ptr, slice) in
+  let list = slice_list_from_auction_state auctions slice.contents.burrow in
+  let _ = if slice_list_is_empty list then
+      failwith "invariant violation: corresponding list for slice was empty"
+    else ()
```

Comment:
Should this be `assert (not (slice_list_is_empty list));`? If we do that it'll be removed when transpiling to ligo.

Reply:
I actually wasn't sure on this point. I tried to mimic the behavior [in liquidationAuction.ml](https://github.com/tezos-checker/checker/blob/ff6259de4d47f7b4a2fe52aaca33e5c6884f4b2c/src/liquidationAuction.ml#L589) since if this case occurs the function will implode.